### DISCUSSION TITLE
CPU Memory Buffer

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#208] [Q&A] [answered]
How would one go about creating a full-CPU-memory-based buffer, such as a vertex buffer that resides entirely in CPU ram? Is this type of thing possible in Diligent Engine?

My D3D9 based engine used these system memory buffers for mesh conversions, and to perform offline operations on them. Its pretty handy, because you can use existing mesh classes during offline operations, while having full read/write access to the data. I've searched around, but could not find information about targeting specific memory pools for buffers.

I think this would relate to D3D12_MEMORY_POOL_L0 vs D3D12_MEMORY_POOL_L1? But I haven't found an equivalent setting in Diligent.

**Comment by TheMostDiligent** [comment, marked as answer]
You will need to implement it yourself. CPU memory is just a CPU memory. When you are done with it, you can copy the data to a GPU (USAGE_DEFAULT) memory buffer.

**Comment by RobertMtx** [comment]
Yeah, I looked at the D3D12 impl of buffer creation and figured that was the case, so I went ahead and wrote a custom buffer class that diverges logic when necessary. I think I was using D3DX9 tools with the buffers in the past, so this wasn't really an option back then. My engine is over 1800 source files and I'm still converting code to Diligent, so no idea yet what trouble I may run into because of the change. Thanks for help!

**Reply by TheMostDiligent** [reply]
Note that if you are re-implementing the same D3D9 concepts in Diligent, you should not expect that it will run faster than D3D9. Probably, slower.
Diligent is a modern graphics API and to leverage its benefits, you need to use modern approaches.

**Reply by RobertMtx** [reply]
Nah, the CPU memory buffers were only being used for offline conversions, such as splitting a mesh into multiple subsets, or mesh editing. They were never used for rendering, or anything GPU based. At least, I don't think they were - I'm still converting through this. For this new situation, I just wrote my own CPU allocator that launches when CPU memory is requested.  Only my native code can see this memory. I haven't modified Diligent - I just wrote a custom wrapper class that makes use of Diligent buffers when GPU memory is used.

Some of the older loading and conversion functions were D3DX9 functions, so I would not have been able to use this approach in the past - DXDX9 would just see an empty buffer when their function was called. But since I won't be relying on D3DX9 anymore, for anything, this shouldn't be a problem. The CPU based meshes are only used for editors and offline conversions (AFAIK), and not during game execution. I can easily convert such a mesh to GPU memory if I needed to render it. In other words, CPU buffers won't have any influence on performance anywhere that it matters. Thanks!
